### PR TITLE
7564 Ordering with JSON attributes

### LINKED ISSUE #7564: Ordering on JSON columns
## Feature
Given the new ability for order by associations, you should also be able to order by JSON for postgres and sqlite.

See https://github.com/sequelize/sequelize/pull/7454#pullrequestreview-30385156

__Dialect:__ postgres / sqlite
__Database version:__ XXX
__Sequelize version:__ 4

### PR DESCRIPTION
**Pull request by mkaufmaner**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

_NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._

### Description of change

Closes #7564 

Added support for ordering by JSON attributes and added escaping to the path to remove a vulnerability that allows SQL injection.

This should make @sushantdhiman happy.

### CODE REVIEW

File: test/integration/model/json.test.js [hunk 1]
```diff
@@ -508,4 +508,192 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
+                first: 'Homer',
+                last: 'Simpson'
+              },
+              employment: 'Nuclear Safety Inspector'
+            });
+          });
+        });
+      });
+
+      it('should query an instance with JSONB data and order', function () {
```
Comment (sushantdhiman) [inline test/integration/model/json.test.js]: Move this test to `find` section above https://github.com/mkaufmaner/sequelize/blob/b5d77ecedbe96d47d053cb9f11e10cdf0f84d0ae/test/integration/model/json.test.js#L209

File: test/integration/model/json.test.js [hunk 2]
```diff
@@ -508,4 +508,192 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
+                first: 'Marge',
+                last: 'Simpson'
+              },
+              employment: 'Housewife'
+            });
+          });
+        });
+      });
+
+      it('should query an instance with JSONB data and order while trying to inject', function () {
```
Comment (sushantdhiman) [inline test/integration/model/json.test.js]: Move this one to find section as well https://github.com/mkaufmaner/sequelize/blob/b5d77ecedbe96d47d053cb9f11e10cdf0f84d0ae/test/integration/model/json.test.js#L209

File: test/integration/model/json.test.js [hunk 3]
```diff
@@ -508,4 +508,192 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
+            field: 'event_data',
+            index: true
+          },
+          json: DataTypes.JSONB
+        });
+
+        return this.Event.sync({ force: true });
+      });
+
+      it('should create an instance with JSONB data', function () {
```
Comment (sushantdhiman) [inline test/integration/model/json.test.js]: This is a redundant test, get rid of it

File: test/integration/model/json.test.js [hunk 4]
```diff
@@ -508,4 +508,192 @@ describe(Support.getTestDialectTeaser('Model'), () => {
...
+                first: 'Homer',
+                last: 'Simpson'
+              },
+              employment: 'Nuclear Safety Inspector'
+            });
+          });
+        });
+      });
+
+      it('should query an instance with JSONB data', function () {
```
Comment (sushantdhiman) [inline test/integration/model/json.test.js]: This test is already covered above, get rid of this as well

File: test/integration/model/json.test.js (hunk 3 above)
Comment (mkaufmaner) [inline test/integration/model/json.test.js]: Removed

File: test/integration/model/json.test.js (hunk 4 above)
Comment (mkaufmaner) [inline test/integration/model/json.test.js]: Removed

File: test/integration/model/json.test.js (hunk 1 above)
Comment (mkaufmaner) [inline test/integration/model/json.test.js]: Moved to find

File: test/integration/model/json.test.js (hunk 2 above)
Comment (mkaufmaner) [inline test/integration/model/json.test.js]: Moved to the SQL injection section below the find.

**Comment by janmeier** [comment]
@mkaufmaner Fails for sqlite, otherwise this is ready to merge

File: lib/dialects/abstract/query-generator.js [hunk 5]
```diff
@@ -795,6 +795,22 @@ const QueryGenerator = {
             } else if (previousModel.rawAttributes !== undefined && previousModel.rawAttributes[item] && item !== previousModel.rawAttributes[item].field) {
               // convert the item attribute from it's alias
               item = previousModel.rawAttributes[item].field;
+            } else if (
+              item.indexOf('.') !== -1
+              && previousModel.rawAttributes !== undefined
+              && previousModel.rawAttributes[item.split('.')[0]].type instanceof DataTypes.JSON
+            ) {
+              // just quote identifiers for now
+              const identifier = this.quoteIdentifiers(previousModel.name  + '.' + previousModel.rawAttributes[item.split('.')[0]].field);
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: `item.split('.')` is called three times in total, that is not good for performance.

Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: Fixed ;)

**Comment by mkaufmaner** [comment]
@janmeier Fixed the sqlite test.

**Pull request merged**